File: 746-min-cost-climbing-stairs/746-min-cost-climbing-stairs.java
class Solution {
    public int minCostClimbingStairs(int[] cost) {
        int dp[]=new int[cost.length+1];
        memset(dp);
     return solveMem(cost,cost.length,dp);   
    }
    void memset(int[]dp){
        for(int i=0;i<dp.length;i++)
            dp[i]=-1;
    }
    int solveMem(int[] arr,int n,int dp[]){

        if(n==0 || n==1)return 0;
        if(dp[n]!=-1)return dp[n];
        int oneStep=solveMem(arr,n-1,dp)+arr[n-1];
        int twoStep=solveMem(arr,n-2,dp)+arr[n-2];
        
        return dp[n]= Math.min(oneStep,twoStep);
    }
    
}

File: 746-min-cost-climbing-stairs/746-min-cost-climbing-stairs.cpp
class Solution {
public:
    int helper(int ind, vector<int>&cost,vector<int>&dp)
    {
        
       if(ind==1 || ind==0)
       {
           return cost[ind];
       }
        if(dp[ind]!=-1)
        {
            return dp[ind];
        }
    return dp[ind]=cost[ind]+min(helper(ind-1,cost,dp),helper(ind-2,cost,dp));
    }
    int minCostClimbingStairs(vector<int>& cost) 
    {
        int n=cost.size();
      vector<int>dp(n+1,-1); 
        return min(helper(n-1,cost,dp),helper(n-2,cost,dp));
    }
};